core/trino-main/src/main/java/io/trino/operator/output/UnnestingPositionsAppender.java

```diff
@@ -94,4 +117,55 @@ private IntArrayList mapPositions(IntArrayList positions, DictionaryBlock block)
         }
         return IntArrayList.wrap(positionArray);
     }
+
+    private static class DictionaryBlockBuilder
+    {
+        private static final int INSTANCE_SIZE = toIntExact(ClassLayout.parseClass(DictionaryBlockBuilder.class).instanceSize());
+        private Block dictionary;
+        private int[] dictionaryIds = new int[0];
+        private int positionCount;
+        private boolean closed;
+
+        public boolean isEmpty()
+        {
+            return positionCount == 0;
+        }
+
+        public Block build()
+        {
+            return DictionaryBlock.create(positionCount, dictionary, dictionaryIds);
+        }
+
+        public long getRetainedSizeInBytes()
+        {
+            return INSTANCE_SIZE + (long) dictionaryIds.length * Integer.BYTES + (dictionary != null ? dictionary.getRetainedSizeInBytes() : 0);
+        }
+
+        public boolean canAppend(Block dictionary)
+        {
+            return !closed && (dictionary == this.dictionary || this.dictionary == null);
```

Comment:
`this.dictionary` could be `null` if `DictionaryBlockBuilder` is not `closed`?

Reply:
yes, this is what happens when `DictionaryBlockBuilder` is created, before the first append